### PR TITLE
Add nil/null support to the cryptString struct

### LINKED ISSUE #4: add nil/null support similar to sql.NullBool
This will allow values to be null:
https://golang.org/pkg/database/sql/#NullBool

### PR DESCRIPTION
**Pull request by sowmiyamuthuraman**
Fixes #4

### CODE REVIEW

File: cryptkeeper.go [hunk 1]
```diff
@@ -14,14 +14,17 @@ import (
 	"reflect"
 )
 
+/* NullString is an alias for sql.NullInt64 data type */
+type NullString sql.NullString
```
Comment (blaskovicz) [inline cryptkeeper.go]: Why did you need to do this?

File: cryptkeeper.go [hunk 2]
```diff
@@ -14,14 +14,17 @@ import (
...
+
 /*CryptString is a wrapper for encrypting and decrypting a string for database operations. */
 type CryptString struct {
 	json.Marshaler
 	json.Unmarshaler
 	sql.Scanner
 	driver.Valuer
 
-	String string
+	String NullString
```
Comment (blaskovicz) [inline cryptkeeper.go]: it may be worth embedding the `NullString` struct rather than making a .String since that will cut out some code noise (ie `.String.String`, `.String.Valid`, etc)

File: cryptkeeper.go [hunk 3]
```diff
@@ -62,6 +68,7 @@ func (cb *CryptBytes) MarshalJSON() ([]byte, error) {
 func (cs *CryptString) UnmarshalJSON(b []byte) error {
 	var target string
 	if err := json.Unmarshal(b, &target); err != nil {
+		cs.String = NullString{Valid: false}
```
Comment (blaskovicz) [inline cryptkeeper.go]: Valid is false by default

File: cryptkeeper_test.go [hunk 4]
```diff
@@ -182,21 +183,21 @@ func TestCryptString(t *testing.T) {
...
 			} else if csError.String != "" {
-				t.Fatalf("Scan of bad encryption should have left String empty, got: %s", csError.String)
+				t.Fatalf("Scan of bad encryption should have left String empty, got: %v", csError.String)
 			}
 
 		})
 		t.Run("Value", func(t *testing.T) {
 			var cs CryptString
-			cs.String = "hello world!"
+			cs.NullString = sql.NullString{"hello world!", true}
```
Comment (blaskovicz) [inline cryptkeeper_test.go]: now that you've embedded the struct, is it possible to do the following:
```suggestion
			cs.String = "hello world!"
			cs.Valid = true
```

File: cryptkeeper_test.go [hunk 5]
```diff
@@ -116,7 +117,8 @@ func TestCryptString(t *testing.T) {
 		if err != nil {
 			t.Fatalf("SetCryptKey should be valid, got: '%s'", err)
 		}
-		cs := CryptString{String: "another secret text"}
```
Comment (blaskovicz) [inline cryptkeeper_test.go:119]: this can stay what it was with, `Valid: true`, similar to how you made the change in the previous code.

File: cryptkeeper_test.go [hunk 6]
```diff
@@ -159,7 +161,7 @@ func TestCryptString(t *testing.T) {
 				t.Fatalf("Scan of string should not have errored: %s", err)
 			}
 			if csString.String != "how are you doing" {
-				t.Fatalf("Scan of string should have matched 'how are you doing', got: '%s'", csString.String)
+				t.Fatalf("Scan of string should have matched 'how are you doing', got: '%v'", csString.String)
```
Comment (blaskovicz) [inline cryptkeeper_test.go]: this can stay as %s

File: cryptkeeper_test.go [hunk 7]
```diff
@@ -182,21 +184,22 @@ func TestCryptString(t *testing.T) {
...
+				t.Fatalf("Scan of int64 should have left String empty, got: %v", csGoofyType.String)
 			}
 
 			var csError CryptString
 			err = csError.Scan(interface{}([]byte("some bad encrypt")))
 			if err == nil {
 				t.Fatalf("Scan of bad encryption should have errored")
 			} else if csError.String != "" {
-				t.Fatalf("Scan of bad encryption should have left String empty, got: %s", csError.String)
+				t.Fatalf("Scan of bad encryption should have left String empty, got: %v", csError.String)
```
Comment (blaskovicz) [inline cryptkeeper_test.go]: same here

File: cryptkeeper_test.go [hunk 8]
```diff
@@ -182,21 +184,22 @@ func TestCryptString(t *testing.T) {
 			if err == nil {
 				t.Fatalf("Scan of int64 should have errored")
 			} else if csGoofyType.String != "" {
-				t.Fatalf("Scan of int64 should have left String empty, got: %s", csGoofyType.String)
+				t.Fatalf("Scan of int64 should have left String empty, got: %v", csGoofyType.String)
```
Comment (blaskovicz) [inline cryptkeeper_test.go]: same here

**Comment by blaskovicz** [comment]
Is it possible to add some test cases for this (specifically for Valid, and also for the null cases) and also add it to CryptBytes as well (since you already added it to CryptString)?

Thank you for your help.

**Comment by sowmiyamuthuraman** [comment]
@blaskovicz for CryptBytes, as there is no structs like sql.NullString for bytes is it fine to introduce the Valid field in the CryptBytes struct.

**Comment by blaskovicz** [comment]
@sowmiyamuthuraman yes
